### DISCUSSION TITLE
[github-actions/github-tags] registryStrategy `hunt` is not taken into account for `getDigest`

### DISCUSSION THREAD

**Discussion by Shegox** [renovatebot/renovate#24162] [Request Help] [answered]
### How are you running Renovate?

self-hosting

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.67.1 as well as main (8f7f941aaffead10e7dcefadae0cf4170b9deb9a)

### If you're self-hosting Renovate, select which platform you are using.

GitHub Enterprise Server (3.8)

### Was this something which used to work for you, and then stopped?

Never saw this working

### Describe the problem

WIP: I will need to investigate a bit further, but wanted to already record my finding in here.

When using Renovate with GitHub Enterprise Servers, Actions can be consumed either from the server itself or directly from github.com. With regards to pull requests #23042 and #23041, Renovate uses the `registryStrategy = 'hunt'` to first search for Actions on the Enterprise Server, and then on github.com if unsuccessful.

The issue arises when Renovate tries to update Actions written in the following format:

```yaml
# filename: .github/workflows/workflow.yaml
      - name: Set up chart-testing
        uses: helm/chart-testing-action@09ed88797198755e5031f25be13da255e7e33aad # v2.3.0
```

Renovate is able to find the update to `v2.4.0`, but fails to resolve the digest. The problem seems to be that the digest is only resolved against the primary data source (i.e., the GitHub Enterprise Server), and does not fall back to the secondary data source (github.com) when needed.

Upon examination of the code, it appears that the digest is determined manually if it's not part of the update:

https://github.com/renovatebot/renovate/blob/36e2b328bba5f7b1a21c3e51dfb98887600e5668/lib/workers/repository/process/lookup/index.ts#L395-L397

A possible solution could be to have the digest reported back directly when querying the GitHub tags. Adding `newDigest: hash,` seems to solve the problem for looking up digests for versioned tags, but an error still occurs when the digest lookup for the current tag (v2.3.0) fails. 
https://github.com/renovatebot/renovate/blob/674f9cbfcc2ba116dbb2d9c96b5333aae8e3d0db/lib/modules/datasource/github-tags/index.ts#L71-L76

---

I'm seeking advice on the best way to implement digest lookup when using `registryStrategy = 'hunt'`? It seems that `github-tags` is the first datasource with registryStrategy hunt and digest support.
 

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Error getting tag commit from GitHub repo (repository=gh-testing/semantic-recommender)
       "githubRepo": "helm/chart-testing-action",
       "err": {
         "message": "Could not resolve to a Repository with the name 'helm/chart-testing-action'.",
         "stack": "Error: Could not resolve to a Repository with the name 'helm/chart-testing-action'.\n    at GithubGraphqlDatasourceFetcher.doRawQuery (/Users/user/github/renovate/lib/util/github/graphql/datasource-fetcher.ts:132:21)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async GithubGraphqlDatasourceFetcher.doShrinkableQuery (/Users/user/github/renovate/lib/util/github/graphql/datasource-fetcher.ts:197:20)\n    at async GithubGraphqlDatasourceFetcher.doPaginatedQuery (/Users/user/github/renovate/lib/util/github/graphql/datasource-fetcher.ts:237:27)\n    at async GithubGraphqlDatasourceFetcher.getItems (/Users/user/github/renovate/lib/util/github/graphql/datasource-fetcher.ts:286:17)\n    at async Function.query (/Users/user/github/renovate/lib/util/github/graphql/datasource-fetcher.ts:57:19)\n    at async queryTags (/Users/user/github/renovate/lib/util/github/graphql/index.ts:15:15)\n    at async GithubTagsDatasource.getReleases (/Users/user/github/renovate/lib/modules/datasource/github-tags/index.ts:70:24)\n    at async getRegistryReleases (/Users/user/github/renovate/lib/modules/datasource/index.ts:75:15)\n    at async huntRegistries (/Users/user/github/renovate/lib/modules/datasource/index.ts:116:13)\n    at async fetchReleases (/Users/user/github/renovate/lib/modules/datasource/index.ts:289:15)\n    at async lookupUpdates (/Users/user/github/renovate/lib/workers/repository/process/lookup/index.ts:86:56)\n    at async withLookupStats (/Users/user/github/renovate/lib/workers/repository/process/fetch.ts:27:18)\n    at async fetchDepUpdates (/Users/user/github/renovate/lib/workers/repository/process/fetch.ts:72:30)\n    at async /Users/user/github/renovate/node_modules/.pnpm/p-map@4.0.0/node_modules/p-map/index.js:57:22"
       }
DEBUG: Could not determine new digest for update. (repository=gh-testing/semantic-recommender)
       "packageName": "helm/chart-testing-action",
       "currentValue": "v2.3.0",
       "datasource": "github-tags",
       "newValue": "v2.3.0",
       "bucket": undefined
DEBUG: packageFiles with updates (repository=gh-testing/some-repo, baseBranch=main)
       "config": {
               {
                 "depName": "helm/chart-testing-action",
                 "commitMessageTopic": "{{{depName}}} action",
                 "datasource": "github-tags",
                 "versioning": "docker",
                 "depType": "action",
                 "replaceString": "helm/chart-testing-action@09ed88797198755e5031f25be13da255e7e33aad # v2.3.0",
                 "autoReplaceStringTemplate": "{{depName}}@{{#if newDigest}}{{newDigest}}{{#if newValue}} # {{newValue}}{{/if}}{{/if}}{{#unless newDigest}}{{newValue}}{{/unless}}",
                 "registryUrls": ["https://github.enterprise.com", "https://github.com"],
                 "currentValue": "v2.3.0",
                 "currentDigest": "09ed88797198755e5031f25be13da255e7e33aad",
                 "updates": [],
                 "packageName": "helm/chart-testing-action",
                 "warnings": [
                   {
                     "message": "Could not determine new digest for update (datasource: github-tags)",
                     "topic": "helm/chart-testing-action"
                   },
                   {
                     "message": "Could not determine new digest for update (datasource: github-tags)",
                     "topic": "helm/chart-testing-action"
                   }
                 ],
                 "sourceUrl": "https://github.com/helm/chart-testing-action",
                 "registryUrl": "https://github.com",
                 "currentVersion": "v2.3.0",
                 "isSingleVersion": true,
                 "fixedVersion": "v2.3.0"
               }
             ],
             "packageFile": ".github/workflows/workflow-helm-test.yaml"
```

</details>

### Have you created a minimal reproduction repository?

Placeholder value, please select the correct response from the dropdown

**Comment by rarkins** [comment]
Ideally the datasource supports/returns a digest with each version, in which case Renovate doesn't need to do any further queries.

Otherwise, it would be best if the `registryUrl` from the version lookup is reused in order to get the digest.

I'm not sure I like the idea of doing the "hunt" over again.

**Reply by Shegox** [reply]
I checked it out and only returning the digest with each version is not sufficient as Renovate needs the digest lookup as well when performing `digest pinning` and at that point the version lookup seems to not be directly. However for some of version lookups this can save us another `getDigest` call

The approach with setting the registryUrl from the version lookup for the digest lookup seems to be working great. 

I created https://github.com/renovatebot/renovate/pull/24175 to implement both.

**Reply by Shegox** [reply, marked as answer]
Resolved with #24175 and #24192. Released in [36.78.5](https://github.com/renovatebot/renovate/releases/tag/36.78.5).
